compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -781,21 +781,28 @@ class Typer extends Namer
     }
   }
 
-  /** For a typed tree `e: T`, if `T` is an abstract type for which an implicit class tag `ctag`
-   *  exists, rewrite to `ctag(e)`.
+  /** For a typed tree `e: T`, if `T` is an abstract type for which an implicit type test or class tag `tt`
+   *  exists, rewrite to `tt(e)`.
    *  @pre We are in pattern-matching mode (Mode.Pattern)
    */
-  def tryWithClassTag(tree: Typed, pt: Type)(using Context): Tree = tree.tpt.tpe.dealias match {
+  def tryWithTypeTest(tree: Typed, pt: Type)(using Context): Tree = tree.tpt.tpe.dealias match {
     case tref: TypeRef if !tref.symbol.isClass && !ctx.isAfterTyper && !(tref =:= pt) =>
-      require(ctx.mode.is(Mode.Pattern))
-      withoutMode(Mode.Pattern)(
-        inferImplicit(defn.ClassTagClass.typeRef.appliedTo(tref), EmptyTree, tree.tpt.span)
-      ) match {
-        case SearchSuccess(clsTag, _, _) =>
-          typed(untpd.Apply(untpd.TypedSplice(clsTag), untpd.TypedSplice(tree.expr)), pt)
-        case _ =>
-          tree
+      def withTag(tpe: Type): Option[Tree] = {
+        require(ctx.mode.is(Mode.Pattern))
+        withoutMode(Mode.Pattern)(
+          inferImplicit(tpe, EmptyTree, tree.tpt.span)
+        ) match
+          case SearchSuccess(clsTag, _, _) =>
+            Some(typed(untpd.Apply(untpd.TypedSplice(clsTag), untpd.TypedSplice(tree.expr)), pt))
+          case _ =>
+            None
       }
+      val tag = withTag(defn.TypeTestClass.typeRef.appliedTo(pt, tref))
+        .orElse(withTag(defn.ClassTagClass.typeRef.appliedTo(tref)))
+        .getOrElse(tree)
+      if tag.symbol.owner == defn.ClassTagClass && !config.Feature.sourceVersion.isAtLeast(config.SourceVersion.`3.1`) then
```

Comment:
That logic seems inverted to me: it says we don't emit any warning about unsound ClassTag usage if we're using 3.1?

Reply:
Oops. Fixed it.